### ISSUE TITLE
Large amount of ram used by MBassador

### ISSUE THREAD

**Issue by dorkbox** [bennidi/eventbus-performance#1]
Here's some initial tests. I also included JVM memory usage, to help understand the data. I don't think I have the disruptor completely worked out, as it seems really, really slow. One thing I did notice is that the default LBQ setting on MBassador used up to 1 GB of ram. Changing it to have an LBQ of size 16 dropped it to about 512 MB without affecting too much. The disruptor with a ring buffer of size 16 uses ~62 MB. I suspect pooling objects would help.

**Linked Blocking Queue (size:Integer.MAX_VALUE)**
![LBQ-intmax-chart](https://cloud.githubusercontent.com/assets/5301521/6149125/d20f15ec-b206-11e4-9526-5b8460f9e72e.jpg)

**Linked Blocking Queue (size:16)**
![LBQ-16-chart](https://cloud.githubusercontent.com/assets/5301521/6149128/d750f3b8-b206-11e4-8ea1-09d9e4365c80.jpg)

**Disruptor (2 worker) + ReflectASM**
![Disruptor-2-worker-pool-reflectasm-chart](https://cloud.githubusercontent.com/assets/5301521/6149140/e7c2cc76-b206-11e4-9716-9f1b10095a24.jpg)

**Comment by bennidi** [comment]
All charts show consistent read performance (Handler Invocation). Write performance varies significantly (high std. deviation) which can be explained by unpredictability in thread scheduling related to synchronized resources. 

Memory consumption is significantly lower for limited queue size but comes with a decrease in performance.

Disruptor is considerably slower (explanation?). Processing is not constant but shows hotspots (clusters of data points). Memory consumption is significantly lower which is partly due to lower throughput.

@dorkbox Any explanations for the low performance of disruptor? Maybe its model does not match the scenario?! Or its not correctly used. What about a custom ring buffer with an atomic index. I already build that once and it wasn't very difficult and gives a constant memory footprint with low synchronization overhead. Maybe an array blocking queue would be a viable alternative, too?!

And many thanks to your great work. It is really interesting to see those figures. I once used JProfiler to trace thread activity (running,waiting,blocked...) to compare MBassador with Guava. I never posted the results (which I should do, I realize). But this could be helpful in figuring out why Disruptor is so much slower. It probably does work that is unnecessary (after all there are not many different consumers/producers).

**Comment by dorkbox** [comment]
I'm not sure why it's so slow - I think that it has to do with queueing messages. I think that the benchmark measure how fast messages can be queued, not necessarily how long it takes to execute each message. 

I'm experimenting with ring buffers and object pools to see how that affects the performance. I must admit I'm rather disappointed in Disruptor's performance. I did read that the primary use for the disruptor is for handling (fixed lengths) bytes of data.

What was interesting, was that having a larger ringBuffer (ie: anything larger than 16) also resulted in progressively worse performance -- and it doesn't make any sense to me why.  think that a larger ring buffer would help queue entries.

**Comment by dorkbox** [comment]
Here's a Multiple Producer, Multiple Consumer queue, instead of the disruptor. It uses a ringbuffer, as you can see by the memory consumption. One of the things I'm looking at is having two queues. One for dispatch, and another for invocation (similar to what you already have).

This is just a single queue + invocation
![mpmcqueue-old-dispatch-chart](https://cloud.githubusercontent.com/assets/5301521/6155900/071fdd64-b236-11e4-8402-d48b1bfe9bcf.jpg)

This is disabling invoking, and just timing the queues (LBQ + MPMC queue).
![dispatch-invoke-only-chart](https://cloud.githubusercontent.com/assets/5301521/6155970/5c4acad8-b236-11e4-9fd1-a67bd570c7b2.jpg)

**Comment by bennidi** [comment]
Hmmm. Why is the Disruptor so slow in comparison? I was so intrigued by their idea, the paper and the many interesting blog posts by Martin Thompson. There must be something wrong in the way it is used.
But I think that the performance of the single queue with ring-buffer is quite impressive: About a million invocations per second. Are the handlers synchronous and do you really measure all their invocations? That would be quite a nice result. Especially when you consider the low memory footprint. What queue implementation are you using?

**Comment by dorkbox** [comment]
Indeed. Their idea, paper, blogs, videos, etc looked really, really good. I _think_ the correct way to use the disruptor is only to hand off data, not to actually process it. I've banged my head on the wall for a while, and I have no idea what I could possibly be doing incorrectly --- so what I've done is moved on to something else.  

I'm using the MPMCqueue from here: http://psy-lob-saw.blogspot.de/2015/01/mpmc-multi-multi-queue-vs-clq.html

Moving to a "dispatch" queue and a "invoke" queue has really helped performance, and using limited queue sizes really helps keep the memory footprint down.

For funsies, here's using a LinkedTransferQueue + MPMC queue.

![2-transfer-chart](https://cloud.githubusercontent.com/assets/5301521/6159285/aa775556-b253-11e4-931f-8ba04906b97b.jpg)

**Comment by dorkbox** [comment]
And here is JUST a LinkedTransferQueue (i'm starting to really like it too)

![only ltq-chart](https://cloud.githubusercontent.com/assets/5301521/6159357/9e9117e4-b254-11e4-8987-43b3f3fe18e1.jpg)

**Comment by bennidi** [comment]
Wow! LinkedTransferQueue is outperforming the rest of the candidates by far in terms of throughput. So, when you need low memory footprint you should take MPMC Queue which will still give good performance (probably best in terms of memory:throughput ratio). If memory is not so much of a concern, then LTQ is the best option.

This is quite an insight. I will definitely bring that into the next release. As for the reflection part: I am still convinced that reflective invocation gets completely optimized away on hotspots and as it is JDK standard I think it is good to keep it the way it is. It is stable and well performing. Do you have other code optimizations/refactorings that are compatible with the current code base that you would like to see in there? Maybe we could collaborate on this project and continue development in on place?

**Comment by dorkbox** [comment]
Based on the performance results, I'm also convinced there is no need for reflectASM -- at least for method access. For field access I have no idea. I'll play around with inflation on the JVM to see what that does too.  

I'll work on getting code into a compatible state the changes based on your master. I did have to make some small changes to the testing framework (adding colors and memory usage, for example), so I'll clean that up and and put in a pull request. It will be a few days though, since I'm swamped with work.

**Comment by dorkbox** [comment]
more info on LTQ performance: http://php.sabscape.com/blog/?p=557

**Comment by dorkbox** [comment]
turns out, i think i figured out the Disruptor. It is REALLY fast and handing off data. really, REALLY fast. The downside, is that if it does any work on that data, I have yet to see it go fast. (as shown earlier)

My test used Disruptor to hand data off to an executor. Wow. The downside, is that the LTQ that it handed data off to grew to about 2 gigs. Not what we want, but interesting.

**Comment by dorkbox** [comment]
I did some benchmarks (notably from [here](https://stackoverflow.com/questions/14146570/calling-a-getter-in-java-though-reflection-whats-the-fastest-way-to-repeatedly/14146919#14146919), because of how misinformed answers are online).

reflective invocation (without setAccessible)   182.837 ns
reflective invocation (with setAccessible)  1.757 ns
reflectASM invocation   0.019 ns
methodhandle invocation 6.391 ns
static final methodhandle invocation    0.019 ns
direct invocation   0.019 ns

**Comment by bennidi** [comment]
Looks like another tweak that can easily be integrated. The good news is that it doesn't require any external library like reflectASM. Only limitation is that it is available only in Java >= 7.

**Comment by dorkbox** [comment]
Maybe, the methodHandle has to be "static final", which means it's tricky. I think reflectASM is the only way to get it down that low, however, 1.7 ns is really, really fast. I'm currently working on some data structures to get object creation down too -- there's a lot going on.

Here's the latest (but not backported yet) -- the timing isn't quite what I want yet, what do you think? It turns out the memory usage isn't quite accurate (it's also measuring the test framework) - I've done some profiling in VisualVM to make improvements

![chart](https://cloud.githubusercontent.com/assets/5301521/6305408/52427eea-b92b-11e4-963d-5439cfdfb463.jpg)

**Comment by dorkbox** [comment]
Here's with subscription & publication concurrent.

1.5x - 2x as fast for read operations.
256 ns/op  VS  169 ns/op

![chart](https://cloud.githubusercontent.com/assets/5301521/6307178/210f907e-b93a-11e4-9ad3-7eb3183307cc.jpg)

**Comment by dorkbox** [comment]
slightly tuned:

~ 138/117 ns/op

![4-chart](https://cloud.githubusercontent.com/assets/5301521/6307661/c9532f0e-b93d-11e4-8724-bfc574472a6b.jpg)

**Comment by bennidi** [comment]
Looks like you are making good progress. How do we go about bringing this back into the core of MBassador? I am currently working on some minor tickets which are mainly about configuration, error handling and stuff. I am also thinking about changing to LTQ and MethodHandle. Have you tweaked other parts?

**Comment by dorkbox** [comment]
I've done some more tweaks (faster iteration over collections) and I'm working on managing memory more efficiently (caching subscriptions for superClasses).  

The only way MethodHandle helps is if it's "static final", which is impossible to do for dynamic method invocation (where JIT will inline the method call).  Reflection via "Method" (as it currently is), or RelflectASM are the only performant options from what I can tell.

Also, I can make the changes to LTQ (i've already got the source included, so it'll work on Java 6) if you'd like.

I'll put in some pull requests in a few days -- I'm finished up some memory issues right now.

**Comment by bennidi** [comment]
Cool. Looking forward to that code. Can you please try and make multiple small PRs? It will help me review the code and understand the changes and their implications. Thanks!

**Comment by dorkbox** [comment]
I'll have it in a lot of commits so you can follow it. Unfortunately, the PR is for the repo (not the commit), so there can't be different PRs for each commit (you just the the whole thing) . :/

**Comment by bennidi** [comment]
The PR is for a specific branch. You could make intermediated branches by
cherry-picking your commits from master such that each branch contains a
closed and working optimization. Otherwise it will be really hard for me. I
can not take a whole bunch of changes, especially the API needs to stay the
same. Please try to sort out your work into meaningful chunks.

Also, the code is quite stable now. There have been no issues with the core
functionality for about a year now. I really want this to continue.

2015-02-22 19:14 GMT+01:00 dorkbox notifications@github.com:

**Comment by dorkbox** [comment]
Of course! I'll put it into separate branches, no problem at all. I agree, keeping it stable is very important.

**Comment by bwzhang2011** [comment]
@dorkbox, how does this going on ? we're looking forward to the boost way for mbassador improvement for queue or disruptor way or some more test as the comparison with guava or rribbit on event bus solution.

**Comment by dorkbox** [comment]
Been super busy with work -- but I've been working on ways to improve the queue/executor, a 0 GC + good performance executor is what I've been working on. A cross between LTQ, Disruptor, and Exchanger. Concurrency is really, really hard - but I'm making solid progress.

**Comment by bennidi** [comment]
@dorkbox Sounds exciting! I am also quite busy with work currently so I will not have much time to spend on mbassador in the next weeks. But I will have a look at your work as soon as you are done, that's for sure. Your participation is greatly appreciated.

**Comment by bwzhang2011** [comment]
@dorkbox , how about your progress for your cross operation for the purpose of improving ?

**Comment by bwzhang2011** [comment]
@dorkbox, any update with such issue  or any idea with your work merged into mbassador branch

**Comment by dorkbox** [comment]
Sorry for taking so long -- I've found some other areas to improve ram/performance (it deals with iterating over certain collections) during my local tests/improvements. I'll be adding (and discussing those) back to the main project)

I'm still testing the executor, and I'll post it as soon as I finish. I'm doing computer science masters-level work (surprisingly few, but really good, papers on this topic) and it's really hard.

**Comment by dorkbox** [comment]
After many months of research and late nights, I've finished the blocking queue -- it's heap allocation is constant (does not change during runtime), and consequently has zero GC; also it scales rather well. The brains of the algorithm are from the EXCELLENT (and ridiculously fast) [MPMC queue written by Nitsan](https://github.com/JCTools/JCTools/tree/master/jctools-core/src/main/java/org/jctools/queues). I re-wrote his MPMC queue to also be a blocking queue (similar to how the LinkedTransferQueue operates). I called it the MpmcTransferArrayQueue (MTAQ), as it is based on the MpmcArrayQueue

I'm still cleaning up/attributing the code, and I'll have it available on github (in my own project), as well as part of MBassador (if approved/wanted).

The performance and memory benefit on my very simplified and structurally changed fork of MBassador was noticable. For MBassador (master), the effects were less, as there is a slight performance improvement - but the more noticeable improvement is in the consumed memory. There are also different ways for MBassador to handle/dispatch messages as well, so that could vary quite a bit.

The following is a performance breakdown of a comparison between LinkedTransferQueue (the latest version is generally accepted as one of the fastest java-collection blocking queues) and MTAQ, each running in different modes and with 2 Consumers/2 Producers, or with 4 Consumers/4 Producers. 

I did not benchmark other configurations, since SingleConsumer/SingleProducer queues would use an entirely different data structure, and most consumer grade hardware will have fewer than 8 cores.

The following tests were on an i7-4700HQ CPU @ 2.40GHz w/ 16gigs of ram, running Linux 3.13.0-49-generic, x86_64.

| Mode (Threads) | LBQ | LTQ | MTAQ |
| --- | --- | --- | --- |
| Blocking (2x2 Threads) | 1,6m op/s | 2,4m op/s | 2,8m op/s |
| Blocking (4x4 Threads) | 1,1m op/s | 1,4m op/s | 2,8m op/s |
| Non-Block (2x2 Threads) | 1,8m op/s | 3,4m op/s | 3,2m op/s |
| Non-Block  (4x4 Threads) | 0,9m op/s | 1,8m op/s | 7,8m op/s |

Given the synchrony of MBassador, I didn't notice an improvement in performance (the dispatch times are only slightly closer to each other, and are likely statistically insignificant), but there is less RAM used since the queue no longer uses the heap.

LinkedBlockingQueue (current master, not LTQ)
![chart](https://cloud.githubusercontent.com/assets/5301521/7460334/7ca8704e-f2a2-11e4-802c-88d168c8a853.jpg)

MTAQ
![chart](https://cloud.githubusercontent.com/assets/5301521/7460340/87ad45dc-f2a2-11e4-881f-e9fb80498cf2.jpg)

(edit: added LinkedBlockingQueue to performance chart)

**Comment by dorkbox** [comment]
For reference, here is a semi-final performance graph in use by my fork of MBassador.

![chart](https://cloud.githubusercontent.com/assets/5301521/7463021/457b5dca-f2b6-11e4-8525-a857bd5e45d3.jpg)

**Comment by bennidi** [comment]
Wow. You have been making quite some progress on the topic. Did I understand correctly that you are doing this work in the context of university studies? I think you are doing a really good job here. What are the dependencies of the code for the MTAQ? Do you use Java classes from JDK 7 or is it Java 6 compatible? If not I think it would be nice to provide them as extensions to the core. It would be too good to have your work become part of the mbassdor project. And can you link the papers you mentioned?

**Comment by bwzhang2011** [comment]
+1 hope such implementation of MTAQ could be provided and as part of the Mbassador.

**Comment by dorkbox** [comment]
I finished years ago at the university - the level of work involved was similar, but my current work is significantly more productive and useful to the community. The dependencies right now are only JCTools (and only a small part of them could be included as source files) and it depends on the use of Unsafe (so java6+ should be ok). I can likely create a version that will only use java's Atomic classes (instead of unsafe), which would make this compatible with JVMs that do not have unsafe access, and android (at the cost of performance). I'll think about how to do that.

Here's the papers i've read on the topic, and the blog by Nitsan (which I recommend reading all of his blogs). To be honest, most of the academic papers were good, but lacking code -- so I cannot prove/reproduce their results. The elimination array, while nice on paper, didn't perform as expected. Maybe with 1024 threads/cores, but not on consumer hardware.

By far, the best was Nitsan's blog/code and LMAX disruptor. Their results _embarrassed_ the papers listed, and there are several masters students/professors that should be ashamed. Nitsan/LMAX products are orders-of-magnitude faster that anything else I could find/think of. Within my specific use-case, Nitsan's MPMC was faster than the disruptor. 

Several observations:
- The disruptor (very good, and amazing performance) falls apart when the queue gets full, as it must then wait for the queue to completely empty before allowing any more data to be added. I think this was the reason I was having performance problems with it.
- The LinkedTransferQueue by Doug Lea, is very, very interesting and really good. There are several versions, and the latest (jdk7) version was ~2x as fast as the original version. The latest version is the one used in the measurement in my previous post.
- Cache-line padding for contented objects makes a **large** difference. [a code example](https://github.com/LMAX-Exchange/disruptor/blob/master/src/main/java/com/lmax/disruptor/util/PaddedLong.java)
- When doing something 70 million times/second
  - Creating new objects is expensive.
  - Casting objects is expensive. (edit: Narrow casting. Wide casting has no overhead)
- The concept of 'Mechanical Sympathy' is brilliant.
- rant: I spent several weeks implementing pseudo-code based on several papers/theisis and was unable to independently verify the results as published. It was incredibly frustrating -- weeks of work, and the results were crap and I suspect the results in their papers were likely made-up. Just because there are 43 pages of theorems proving your approach is faster/better/correct, doesn't mean it actually is. Well, it might be correct, but it's definitely not fast. It is an academic embarrassment that Computer Science papers don't have mandatory inclusion for source code. We don't use punch-cards anymore -- source code is trivial to include with the published paper. /rant

Academic papers
- [The Baskets Queue](https://people.csail.mit.edu/shanir/publications/Baskets%20Queue.pdf)
- [Efficient CAS](http://www.cs.bgu.ac.il/~hendlerd/papers/EfficientCAS.pdf)
- [A Dynamic Elimination-Combining Stack Algorithm](http://www.cs.bgu.ac.il/~hendlerd/papers/DECS.pdf)
- [Cache-Aware Lock-Free Queues for Multiple
  Producers/Consumers and Weak Memory Consistency](http://www.cse.chalmers.se/~tsigas/papers/Cache-Aware-Queues-OPODIS10.pdf)
- [Fast Concurrent Queues for x86 systems](http://www.cs.technion.ac.il/~mad/publications/ppopp2013-x86queues.pdf)
- I'm missing a few, but these were the ones that I had bookmarked.

And Nitsan's blog
- [Notes On Concurrent Ring Buffer Queue Mechanics](http://psy-lob-saw.blogspot.de/2014/04/notes-on-concurrent-ring-buffer-queue.html)
- [MPMC: The Multi Multi Queue vs. CLQ](http://psy-lob-saw.blogspot.de/2015/01/mpmc-multi-multi-queue-vs-clq.html)
- [The code repo](https://github.com/JCTools/JCTools/tree/master/jctools-core/src/main/java/org/jctools/queues)

Other good blogs
- [JcTools based queues performance](http://fasterjava.blogspot.de/2014_09_01_archive.html)
- [A Pair of (somebody else’s) Concurrency Bugs](http://www.azulsystems.com/blog/cliff/2011-09-23-a-pair-of-somebody-elses-concurrency-bugs)
- [Concurrent Sequencing](http://bad-concurrency.blogspot.de/2011/12/concurrent-sequencing_31.html)

Other good code references
- [Caffeine](https://github.com/ben-manes/caffeine/tree/master/caffeine/src/main/java/com/github/benmanes/caffeine)

**Comment by dorkbox** [comment]
Added LinkedBlockingQueue to performance chart.

| Mode (Threads) | LBQ | LTQ | MTAQ |
| --- | --- | --- | --- |
| Blocking (2x2 Threads) | 1,6m op/s | 2,4m op/s | 2,8m op/s |
| Blocking (4x4 Threads) | 1,1m op/s | 1,4m op/s | 2,8m op/s |
| Non-Block (2x2 Threads) | 1,8m op/s | 3,4m op/s | 3,2m op/s |
| Non-Block  (4x4 Threads) | 0,9m op/s | 1,8m op/s | 7,8m op/s |

**Comment by bennidi** [comment]
It seems to me that you did an awesome job with your research and implementations. What are your plans to make all this available to the community? Is your fork ready for production? I could link it from the README and you maybe could write up some documentation that reflects your insights and put them in the wiki of your fork. I am also thinking about providing the different queue implementations as extension. Unfortunately, I am super busy working on another project of mine and do not know when I will have a day or two to work on that matter.

I also need to address an issue related to different ordering of reflected methods reported in https://github.com/bennidi/mbassador/issues/101

**Comment by dorkbox** [comment]
I plan on contributing it back to jctools, hopefully it'll be accepted -- otherwise I'll just host it on github, and of course, make a version avilable to mbassador. There are two basic versions, and it involes different tweaks necessary to fully implement the TransferQueue interface.

My fork isn't quite ready for production, mostly because the non-blocking performance can be significantly better (ex: 70m op/s is baseline MpmcArrayQueue performance). The history is also mess -- so much junk in it from all the different experimentation, so I'll have to rework things so it can have a clean history. I'm pretty busy with work and side projects, and I suspect I'll get to it soon.

**Comment by bennidi** [comment]
Looking forward to that. I am sure the JCTools people will like your work. It really seems like you have thoroughly researched the relevant issues. I am quite impressed definitely. And I am very glad that one of my libraries was remotely involved in that endeavour.

**Comment by dorkbox** [comment]
I'll consolidate the tests as well, since some of the numbers for MTAQ look really good, and I hope it's not because of typos in the (late night) test harness/code.  I will re-run/post the test results once I have things in a cleaner state.

**Comment by bwzhang2011** [comment]
@dorkbox, hope you excellent work could be merged into mbassador as the default one or plugin one. maybe it's a good idea to submit them once it satisfied most of your requirements and tried to improve it for the following versions.

**Comment by dorkbox** [comment]
Updated performance chart. Things have settled, and I am now getting more consistent results.

| Blocking | LBQ | LTQ | MTAQ |
| --- | --- | --- | --- |
| 1x1 Threads | 3,7m op/s | 6,3m op/s | 3,6m op/s |
| 2x2 Threads | 1,9m op/s | 2,4m op/s | 2,9m op/s |
| 3x3 Threads | 1,4m op/s | 1,8m op/s | 2,0m op/s |
| 4x4 Threads | 0,8m op/s | 1,4m op/s | 1,5m op/s |

| Non-Blocking | LBQ | LTQ | MTAQ |
| --- | --- | --- | --- |
| 1x1 Threads | 5,8m op/s | 8,1m op/s | 30,7m op/s |
| 2x2 Threads | 1,4m op/s | 3,4m op/s | 6,4m op/s |
| 3x3 Threads | 1,2m op/s | 2,3m op/s | 3,8m op/s |
| 4x4 Threads | 0,8m op/s | 1,8m op/s | 2,4m op/s |

**Comment by bwzhang2011** [comment]
@dorkbox, any update with your issue. as your testing cases, you take jctools api into consideration for testing. but when I check out the resource codes I found it required us to use jdk8 for building. my question is whether you would suggest using jctools for high performance or extract the MTAQ implementation into mbssador which make it JDK7 compatible.

**Comment by CodeMason** [comment]
Only small updates -- I'm working through and evaluating collection performance for the different collections used by my fork and by MBassador. I've played with these in the past, but have not found consistent benchmarks just for the collections, nor any meaningful implementation that improved performance.

The LBQ/LTQ/MTAQ benchmarks themselves don't really depend on JCTools, but MTAQ does (as it depends on MpmcArrayQueue). Right now Nitsan is in the middle of a code review.


Running jctools is JDK6, building is JDK8. I'll setup a repo for all the tests I have, so you can check them out, and I'll include all the libraries as well, so running things should be straight forward.

**Comment by dorkbox** [comment]
Only small updates -- I'm working through and evaluating collection performance for the different collections used by my fork and by MBassador. I've played with these in the past, but have not found consistent benchmarks just for the collections, nor any meaningful implementation that improved performance.

The LBQ/LTQ/MTAQ benchmarks themselves don't really depend on JCTools, but MTAQ does (as it depends on MpmcArrayQueue). Right now Nitsan is in the middle of a code review.

Running jctools is JDK6, building is JDK8. I'll setup a repo for all the tests I have, so you can check them out, and I'll include all the libraries as well, so running things should be straight forward.

EDIT: whoops, project tracking account was accidentally used.

**Comment by bwzhang2011** [comment]
@dorkbox, thanks for view and explanation and jctools with small but helpful high performance collections is much popular and appreciated. hope such could be provided not long in the future and as the official support in mbassador. thanks for your great efforts it seems that you're more strict with your work and  most of us prefer simple but working idea so that's the difference and distance.

**Comment by bennidi** [comment]
@dorkbox Any news from the code reviewing. I currently have some time to work on MBassador again.

**Comment by dorkbox** [comment]
some news. I have a question though: What's the target JDK version for running (and/or compiling)? 

(and an update: I will be posting the test code at some point, I've just been really busy.)

I've found good improvements using arrays with array-iteration (instead of for-each, which has more byte-code instructions). The use of arrays was significantly more performant than trying to use  concurrent collections, or collections concurrently (it was a valiant effort to try to use concurrent collections, but they were just too slow.). The difference was about 100-250ms per 1000-publication-batch. Caching the super classes and their subscriptions (and clearing the super-subscriptions on sub/unsub) also helps a lot with super-subscriptions.

_Based on the experiments I've done, I'm satisfied that this is the absolute fastest publications can happen._

This is in a separate branch (lots, and lots of experimenting/changing code), but what ultimately was done, was to store everything as ArrayList, and use .toArray() to return the subscriptions. Combining ArrayLists is also a lot faster than combining other data structures, because a system call to copy the backing array is used (instead of iterating over each element to add it). Since these are relatively small arrays, it's extremely fast.

The other change is that I'm currently using JDK8 StampedLock (instead of RWLocks), and the benefit is noticeable, but small.

I found the best java (that produced bytecode without casting the array) was the following:

``` java
if (collection != null) {
    final Subscription[] subscriptions = new Subscription[collection.size()];
    collection.toArray(subscriptions);
    return subscriptions;
}
else {
   return null;
}
```

Here's my current WIP and comparison (Mbassador is first). I had to increase the BatchesPerPublisher to 20, and the SubscriberDelay to 0 in order to capture the results. These charts are for the Read Write High Concurrency test.

![chart-mbassador](https://cloud.githubusercontent.com/assets/2230263/8205709/4ba3090a-14f3-11e5-9574-a3b60428eeae.jpg)

![chart](https://cloud.githubusercontent.com/assets/2230263/8205732/9d95f600-14f3-11e5-8392-bead12a5263c.jpg)

**Comment by bennidi** [comment]
So there are quite clear performance gains. Can you chunk the most important parts into smaller PRs again?
I am also thinking about how to make components like the MTAQ be loaded as additional dependency instead of including it into the core. There might be other implementations targeting specific platforms which could be made available with a sort of "runtime discovery" mechanism (looking up class by name or something). Don't know if this would be the right way to go. Have you thought about publishing MTAQ as a standalone library?

**Comment by dorkbox** [comment]
Sure, I'll work on pieces and submit PRs.

I'm still waiting on Nitsan for code review. It's a tool i'll be using, so it'll be published standalone with my other tools once I get it into my build system and I hear back from Nitsan. While fast and slim, I don't want to prematurely publish it, since there are many mistakes I could have made. I spent a long time reviewing it, so I hope not.

**Comment by roeltje25** [comment]
Any news on integrating the performance improvements into the MBassador master branch?

**Comment by bwzhang2011** [comment]
@dorkbox, any update with the issue and when could we take a look at such in mbassador master branch compared with jctools.

**Comment by dorkbox** [comment]
jctools did a code review about a week ago -- but I've been swamped at work. We have a release next week, and I'll be able to finish the code-review and array enhancements then.

**Comment by bwzhang2011** [comment]
@dorkbox, that's a good news and thanks a lot for your great efforts to improve the mbassador performance.

**Comment by roeltje25** [comment]
Great I'm looking forward to it.

**Comment by dorkbox** [comment]
update: My companies release got pushed back 2 weeks, so there's been that... I'll work on merging my changes in this weekend.

**Comment by bwzhang2011** [comment]
@dorkbox, any update with such issue ?

**Comment by CodeMason** [comment]
Yes. Going to be submitting code later today

**Comment by bwzhang2011** [comment]
@dorkbox, any update with such issue ?

**Comment by dorkbox** [comment]
@bwzhang2011 https://github.com/bennidi/mbassador/pull/125

**Comment by bwzhang2011** [comment]
@dorkbox, thanks for following such issue and make further improvement with mbassador.

**Comment by dorkbox** [comment]
There are a few slight improvements (once I finish the MTAQ review for JCTools), so I'm leaving this issue open until that is complete.

**Comment by bwzhang2011** [comment]
Thanks for review. as new mbassador released with bennidi/mbassador#125 brought it, looking forwar d to your MTAQ modification. for another side,  I want to do some integration with axon dispatch commandbus and mbassador integration and I think mbassador could leave use huge performance once MTAQ is merged.

**Comment by bwzhang2011** [comment]
@dorkbox, any update with such issue ?

**Comment by dorkbox** [comment]
I'm busy finishing a fork/fix/rewrite of the Universal TweenEngine (https://github.com/dorkbox/TweenEngine) - as it had some nasty bugs coupled with a really complicated state machine, and I don't want to context switch until it's done (which hopefully is soon. It's tricky, especially concerning things like GC, reducing memory usage and trimming unnecessary calculations.

**Comment by dorkbox** [comment]
@bwzhang2011 I have updated the pull request with JCTools, and once it's merged - I will finish the implementation of MTAQ for mbassador.

**Comment by bwzhang2011** [comment]
@dorkbox, thanks a lot for your great efforts for performance improvement for mbassador. I will continue to add such in my project. what's more, it will be better take the distribute way into consideration. maybe in the future @bennidi  should take some way for that to be the orientation. as jctools bring in some IPC way, maybe mbassador could develop some other project for that.

**Comment by roeltje25** [comment]
@dorkbox 
It's been some time since you mentioned finishing the MTAQ for mbassador. What is the status? I am also concerned about continuing progress on mbassador now @bennidi has mentioned that he will not continue to actively develop mbassador.

anyway, I am interested to see performance of mbassador increase, as it's the core of our development here. A big thanks to @bennidi sofar

**Comment by dorkbox** [comment]
@roeltje25 
[Waiting for my pull request to be merged](https://github.com/JCTools/JCTools/pull/68). Life got in the way (both for nitsanw and myself), and hopefully my changed will get merged soon.

I wouldn't be concerned about mbassador, as it is WAY better than other known message bus implementations. It's stable and incredibly simple (as things go) which is why @bennidi doesn't need to actively develop it. 

I should also mention: mbassador is very fast on it's own, and the _only_ technique I could discover to dramatically improve it's performance was to strip out a lot of features; almost entirely to do with object creation. The performance improvement that MTAQ brings, is that it is based on insanely fast queues (from JCTools) and removes object creation. The backbone of (pretty much all) thread executors are queues, and MTAQ addresses this. 

[I have a fork](https://github.com/dorkbox/MessageBus) that I'm using internally, which is a stripped down version of mbassador, but I wouldn't use it yet -- I'm waiting on the final version of JCTools and then another set of regression testing before recommending it in production systems.

**Comment by nitsanw** [comment]
@dorkbox @roeltje25  indeed I have been slow to review the PR :-( I will make it a priority

**Comment by bwzhang2011** [comment]
@nitsanw, any update with such issue epecially with the PR from dockbox ?

**Comment by nitsanw** [comment]
@bwzhang2011 I have reviewed the PR, see dialogue here: https://github.com/JCTools/JCTools/pull/68
The bottom line is, I have made some corrections/suggestions to the original impl which @dorkbox accepted. I cannot accept the PR at this time as it does not fully implement TransferQueue, but within limitations the implementation is correct and beneficial. If the limited functionality it offers is sufficient for your need than I would suggest you start with that.
@dorkbox please correct me if I'm off the mark.
Thanks

**Comment by dorkbox** [comment]
@nitsanw  is correct, and I'm currently investigating what to do for the purposes of mbassador.

I should have an update on this issue soon, as my work schedule permits.

**Comment by bwzhang2011** [comment]
@dorkbox, esp for myself much appreciated of your attitude for your great work and experiment efforts for such data structure and hope your implementation could be integrated with mbassador no matter whether it would be fully accepted with jctools.

**Comment by dorkbox** [comment]
The architecture of mbassador just won't work with some of the enhancements that I have identified. As a result, I have forked (and changed some functionality) of mbassador to accomplish this. It's a _much_ [simpler version of mbassador](https://github.com/dorkbox/MessageBus), but is a bit faster as a result of **fewer** features along with my enhancements. I would say it is more of a "pure" pub/sub messagebus, with no frills (sorting, priority, filters, etc), it's just subscribe and publish with _high-performance_ async publication.

On that topic, and a rather important note: the use of the disruptor for async publication is **significantly** faster than anything else. Synchronous publication is a little bit faster, and outside of benchmarks I don't think it would be noticeable.

All is not lost, as I will be issuing a PR for adding a limited, but appropriate, enhancements to mbassador. _Just to make extremely clear_ ... MBassador is _already_ really fast, and there's just not a whole lot to make faster. Specifically, this PR will be implementing the `single-writer-principle` outlined by [Nitsan Wakart](http://psy-lob-saw.blogspot.de/2012/12/atomiclazyset-is-performance-win-for.html).

I will attach benchmarks in the next two posts.

**Comment by dorkbox** [comment]
This is **synchronous** publication. The first is _MBassador_, the second is my [fork](https://github.com/dorkbox/MessageBus).
![MBassador](https://cloud.githubusercontent.com/assets/5301521/13724312/b5028b20-e881-11e5-8a65-4f6eb0316307.jpg)

![MessageBus](https://cloud.githubusercontent.com/assets/5301521/13724314/b8e56280-e881-11e5-87fe-0357681f7dde.jpg)

**Comment by dorkbox** [comment]
This is **asynchronous** publication. The first is _MBassador_, the second is my [fork](https://github.com/dorkbox/MessageBus).

![1452812619697_mbassador-async](https://cloud.githubusercontent.com/assets/5301521/13724321/f72bbe54-e881-11e5-904b-d7fc5d62b716.jpg)

![1457803804204_messagebus_async](https://cloud.githubusercontent.com/assets/5301521/13724326/05531680-e882-11e5-90cd-b01a7130f712.jpg)

**Comment by dorkbox** [comment]
For more detailed and extensive tests, see my [Benchmarks project](https://github.com/dorkbox/Benchmarks/tree/master/src/dorkbox/benchmark)  

These tests aren't to describe "real world" performance, but to derive comparisons between different implementations, and because of
something called OSR, you **CANNOT** depend on these tests to describe what they are testing in the "real world".

**Comment by nitsanw** [comment]
@dorkbox Fair attribution: the single-writer-principle is a @mjpt777 term I have used, but not originated. See the blog post here: http://mechanical-sympathy.blogspot.co.za/2011/09/single-writer-principle.html

**Comment by bennidi** [comment]
@dorkbox I see that you managed to gain around 25% of performance in synchronous dispatch (MBassador ~250ms and your fork ~ 180ms for two million handlers). That is impressive. I would not have expected this margin to be available :) But as you said, you had to remove some of the features that I would consider an important part of the library. Anyways, it's great to see that your efforts were rewarded. I will gladly include a reference to your project in MBassadors main readme.

As to the graphs about async dispatch, I think I am not able to interpret them correctly as the look completely different. What do you get out of them? I would also be very happy to hear a brief summary of your learnings with Disruptor. If I remember correctly you were struggling in the beginning to make it work. What were you doing wrong?

Thanks again for all your work.

**Comment by dorkbox** [comment]
@bennidi You're very welcome, and thank you for the mention. This has been an interesting _journey_, and I have learned an incredible amount about concurrent programming, what works and what doesn't. Also, I'm really happy with my results, and it _feels great_ to claim <100ns per message dispatched.

The **major** performance contribution was applying the `single-writer-principle` (I recommend reading the this blog post for details http://mechanical-sympathy.blogspot.co.za/2011/09/single-writer-principle.html). 

The other optimizations (in order of how much they enhanced the performance) were to remove as many branch conditions as possible -- which had the side affect of removing a bunch of features; to use faster collections (the Kryo [IdentitiyMap](https://github.com/EsotericSoftware/kryo/blob/master/src/com/esotericsoftware/kryo/util/IdentityMap.java) instead of HashMap); and to modify the use of your ConcurrentSet Iterators ([Strong](https://github.com/bennidi/mbassador/blob/master/src/main/java/net/engio/mbassy/common/StrongConcurrentSet.java#L21) and [Weak](https://github.com/bennidi/mbassador/blob/master/src/main/java/net/engio/mbassy/common/WeakConcurrentSet.java#L26)) to not generate objects - ([here](https://github.com/dorkbox/MessageBus/blob/master/src/dorkbox/messagebus/subscription/reflection/SubscriptionReflectionStrong.java#L95) and [here](https://github.com/dorkbox/MessageBus/blob/master/src/dorkbox/messagebus/subscription/reflection/SubscriptionReflectionWeak.java#L116)). Removing object generation was a goal of mine, and I'm not convinced that change had much of an impact on the performance... If the charts were a _straight line_ it would be easier to measure.

You'll see that anywhere there is concurrent access to a collection, the `single-writer-principle` is applied -- and for some areas, it just wouldn't be easy to apply to MBassador without changing the architecture. I think the main area it could be applied (and this would be what my pull request would be) is to replace the re-entrant locks in the SubscriptionManager and HashMap with IdentityMap.

> RE: Async dispatch

Those graphs are a bit tricky to interpret for me as well. The _best_ I can think of to explain the differences... Since MBassador queues all of the publications (which is what originally started me down this "rabbit hole") the tests finish running with nothing/very little actually getting published. The difference is that via the Disruptor, it's fast enough to be able to "stay on top" of processing the publications.

> RE: The LMAX-Disruptor.

Yes - I was really, really struggling with the Disruptor - this was a year ago and so my memory is a little fuzzy on the details, so I will explain how I got it to work in _round 2_.  I don't know the exact problem a year ago, but I tried to modify the example and use "it all at once", and it failed to perform; and then again on my first "retry" a few months ago, I modified it all at once - and it failed to perform. My success came when I took a [performance test example](https://github.com/LMAX-Exchange/disruptor/blob/master/src/perftest/java/com/lmax/disruptor/workhandler/TwoToTwoWorkProcessorThroughputTest.java) and **very slowly** adapted it ([while benchmarking after each modification](https://github.com/dorkbox/Benchmarks/blob/master/src/dorkbox/benchmark/Test_Queue_Disruptor.java)) to my own use.  The best I can say, is that the Disruptor is _extremely sensitive_ to all of it's parameters, and any changes made that differ from the examples have to be benchmarked to make sure those changes didn't break anything in the process.

**Comment by nitsanw** [comment]
@dorkbox would a single writer Identity/Hash Map/Set help here?

**Comment by dorkbox** [comment]
@nitsanw Currently the `put()` and `get()` are wrapped to use the single-writer-principle -- is that what you mean?

**Comment by cklsoft** [comment]
Great job.

**Comment by nitsanw** [comment]
@dorkbox I mean a single thread updates the map/set. Multi/single reader makes no odds in the map/set case.
